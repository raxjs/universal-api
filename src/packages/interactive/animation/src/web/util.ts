import { AnimationAction, AnimationOptions } from '../types';
import { normalizeOptions } from '../common';
import { CONTAINER_NAME } from '@utils/constant';

/**
 * normalize unit
 * @param value
 * @param unit
 * @param force
 */
export function normalizeUnit(value: number | string, unit: string, force = false): string {
  if (force || typeof value === 'number') {
    return `${value}${unit}`;
  }
  return value;
}

/**
 * merge default options
 * @param options
 */
export function getDefaultOptions(options?: AnimationOptions): AnimationOptions {
  return normalizeOptions(
    {
      duration: 400,
      timingFunction: 'linear',
      delay: 0,
      transformOrigin: '50% 50% 0',
      ...options,
    },
    CONTAINER_NAME.WEB,
  );
}

/**
 * handle animation actions queue
 * @param actions
 * @param fn
 * @param onFinish
 */
export function handleActionsQueue(
  actions: AnimationAction[],
  fn: (action: AnimationAction, callback: () => void) => any,
  onFinish?: () => any,
) {
  actions = actions.slice();
  const action = actions.shift();
  if (action) {
    fn(action, () => handleActionsQueue(actions, fn, onFinish));
  } else if (onFinish) {
    onFinish();
  }
}
