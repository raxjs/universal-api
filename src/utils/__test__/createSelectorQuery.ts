type ContextMethods =
  'createSelectorQuery' | 'select' | 'selectAll' | 'selectViewport' | 'in' | 'exec'
  | 'boundingClientRect' | 'context' | 'getContext' | 'fields' | 'node' | 'scrollOffset'
  | string;

export class MockSelectorQueryImpl {
  selectorQueryContext: Record<ContextMethods, jest.Mock>;
  nodesRefContext: Record<ContextMethods, jest.Mock>;
  execResult: any;

  constructor(result?: any) {
    this.selectorQueryContext = {};
    this.nodesRefContext = {};
    this.setExecResult(result);
    this.initContext();
  }

  set(
    method: ContextMethods,
    contextType: 'SelectorQuery' | 'NodesRef',
    returnedType: 'SelectorQuery' | 'NodesRef',
    impl?: (...args: any[]) => void,
  ) {
    const ctx = contextType === 'SelectorQuery'
      ? this.selectorQueryContext
      : this.nodesRefContext;
    ctx[method] = jest.fn((...args: any[]) => {
      if (impl) {
        impl(...args);
      }
      return returnedType === 'SelectorQuery'
        ? this.selectorQueryContext
        : this.nodesRefContext;
    });
    return this.selectorQueryContext[method];
  }

  setExecResult(result?: any) {
    this.execResult = result;
    this.getMock('exec')?.mockClear();
    this.set('exec', 'SelectorQuery', 'NodesRef', (cb) => {
      cb?.(this.execResult);
    });
  }


  getMock(method: ContextMethods): jest.Mock | undefined {
    let res = this.selectorQueryContext[method];
    if (!res) {
      res = this.nodesRefContext[method];
    }
    return res;
  }

  clearMock() {
    Object.keys(this.selectorQueryContext).forEach((method) => {
      this.getMock(method).mockClear();
    });
    Object.keys(this.nodesRefContext).forEach((method) => {
      this.getMock(method).mockClear();
    });
  }

  clear() {
    this.clearMock();
    this.initContext();
  }

  private initContext() {
    this.selectorQueryContext = {};
    this.nodesRefContext = {};

    this.set('createSelectorQuery', 'SelectorQuery', 'SelectorQuery');
    this.set('in', 'SelectorQuery', 'SelectorQuery');
    this.set('select', 'SelectorQuery', 'NodesRef');
    this.set('selectAll', 'SelectorQuery', 'NodesRef');
    this.set('selectViewport', 'SelectorQuery', 'NodesRef');
    this.set('exec', 'SelectorQuery', 'NodesRef', (cb) => {
      cb?.(this.execResult);
    });

    this.set('boundingClientRect', 'NodesRef', 'SelectorQuery');
    this.set('context', 'NodesRef', 'SelectorQuery');
    this.set('getContext', 'NodesRef', 'SelectorQuery'); // bytedance
    this.set('fields', 'NodesRef', 'SelectorQuery');
    this.set('node', 'NodesRef', 'SelectorQuery');
    this.set('scrollOffset', 'NodesRef', 'SelectorQuery');
  }
}

