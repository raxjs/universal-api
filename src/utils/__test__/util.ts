import { JSDOM } from 'jsdom';

// Note: `ali` includes: `dingtalk`
type Container = 'web' | 'wechat' | 'ali' | 'dingtalk' | 'bytedance' | 'kuaishou' | 'baidu';
type Globals = Record<'window' | 'wx' | 'my' | 'dd' | 'tt' | 'ks' | 'swan' | string, any>;

export const mapContainerToGlobalsKey: Record<Container, keyof Globals> = {
  web: 'window',
  wechat: 'wx',
  ali: 'my',
  dingtalk: 'dd',
  bytedance: 'tt',
  kuaishou: 'ks',
  baidu: 'swan',
};

export function createNoop() {
  return () => {
  };
}

export const noop = createNoop();

export function sleep(ms: number) {
  return new Promise((resolve) => {
    setTimeout(resolve, ms);
  });
}

export function createPromisifyImpl(value: any = {}) {
  return (args) => {
    setTimeout(() => {
      args?.success?.(value);
    }, 0);
  };
}

export function isAliContainer(container: Container) {
  return container === 'ali' || container === 'dingtalk';
}

/**
 * run test in mock container environment for special api
 * @param container
 * @param callback
 * @example see example at: src/packages/interactive/animation/__test__/api.test.ts
 */
export function testContainerAPI(
  container: Container,
  callback: (globals: Partial<Globals>) => any,
) {
  const map: Record<Container, Globals> = {
    web: {},
    wechat: { wx: { request: noop } },
    ali: { my: { alert: noop } },
    dingtalk: { my: { alert: noop }, dd: { alert: noop } },
    bytedance: { tt: { showToast: noop } },
    kuaishou: { ks: { showToast: noop } },
    baidu: { swan: { showToast: noop } },
  };

  if (container === 'web') {
    const { window } = new JSDOM();
    map.web.window = window;
    map.web.document = window.document;

    // innerText implementation
    // @see https://github.com/jsdom/jsdom/issues/1245
    Object.defineProperty(window.HTMLElement.prototype, 'innerText', {
      get() {
        return this.textContent;
      },
      set(val: string) {
        this.innerHTML = val.replace(/</g, '&lt;').replace(/>/g, '&gt;');
      },
    });
  }

  const globals = map[container] || {};

  const proxyGlobals = new Proxy({}, {
    set(target: Globals | {}, p: PropertyKey, value: any, receiver: any): boolean {
      global[p] = value;
      return Reflect.set(target, p, value, receiver);
    },
    defineProperty(target: Globals | {}, p: PropertyKey, attributes: PropertyDescriptor): boolean {
      Object.defineProperty(global, p, attributes);
      return Reflect.defineProperty(target, p, attributes);
    },
    deleteProperty(target: Globals | {}, p: PropertyKey): boolean {
      delete global[p];
      return Reflect.deleteProperty(target, p);
    },
  });
  test(`Test container: ${container}`, async () => {
    // 执行测试前清除 require 模块缓存
    jest.resetModules();

    // set
    Object.assign(proxyGlobals, globals);

    try {
      // callback
      await callback(proxyGlobals);
    } finally {
      // clear
      Object.keys(proxyGlobals).forEach((k) => {
        delete global[k];
      });
    }
  });
}

type ConfigAPI = (api: string, mockImpl: jest.Mock) => void;

/**
 * batch run test in multiple mock container environment for special api
 * @param name
 * @param containers
 * @param callback
 */
export function testPlatformAPI(
  name: string,
  containers: Container[],
  callback: (container: Container, globals: Partial<Globals>, configAPI: ConfigAPI) => any,
) {
  describe(`Test Platform API: ${name}`, () => {
    for (const container of containers) {
      testContainerAPI(container, async (globals) => {
        const configAPI = (api: string, mockImpl: jest.Mock) => {
          const key = mapContainerToGlobalsKey[container];
          if (globals[key]) {
            globals[key][api] = mockImpl;
          }
        };
        await callback(container, globals, configAPI);
      });
    }
  });
}


export function testWebAPI(
  name: string,
  callback: (globals: Partial<Globals>) => any,
) {
  describe(`Test Platform API: ${name}`, () => {
    testContainerAPI('web', async (globals) => {
      await callback(globals);
    });
  });
}
