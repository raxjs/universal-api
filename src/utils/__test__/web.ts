type TriggerType = 'click' | 'focus' |'blur' | 'change' ;

export class DOMUtil {
  element: Element;

  constructor(ele: string | Element, index = 0) {
    if (typeof ele === 'string') {
      const nodeList = document.querySelectorAll(ele);
      this.element = nodeList[index] ?? null;
    } else {
      this.element = ele;
    }
  }

  isAppeared() {
    if (!this.element) return false;
    return this.getComputedStyle()?.display !== 'none';
  }


  /**
   * get element computed style
   */
  getComputedStyle() {
    if (this.element) {
      return window.getComputedStyle(this.element);
    }
    return null;
  }

  toString() {
    if (this.element) {
      return this.element.innerHTML;
    }
    return '';
  }

  trigger(type: 'click' | 'focus' | 'blur'): void;
  trigger(type: 'change', value: string | number): void;

  /**
   * trigger dom event
   * @param type
   * @param arg
   */
  trigger(type: TriggerType, arg?: any) {
    let event: Event;
    switch (type) {
      case 'click':
        event = new window.MouseEvent(type);
        break;
      case 'focus':
      case 'blur':
        event = new window.FocusEvent(type);
        break;
      case 'change':
        event = new window.Event(type);
        (this.element as HTMLInputElement).value = String(arg);
        break;
      default:
    }

    if (this.element && event) {
      this.element.dispatchEvent(event);
    }
  }
}
