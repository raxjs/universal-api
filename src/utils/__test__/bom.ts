import { sleep } from './util';
import { DOMUtil } from './web';

/**
 * 模拟 localStorage
 * @param globals
 */
export function mockLocalStorage(globals) {
  const localStorage = {
    _data: new Map<string, string>(),
    setItem: (key: string, value: any) => {
      localStorage._data.set(key, String(value));
    },
    getItem: (key: string): string => {
      return localStorage._data.get(key) ?? null;
    },
    removeItem: (key: string) => {
      localStorage._data.delete(key);
    },
    clear: () => {
      localStorage._data.clear();
    },
  };

  globals.localStorage = localStorage;
  Object.defineProperty(window, 'localStorage', {
    get() {
      return localStorage;
    },
  });
}

/**
 * 模拟 xhr 对象
 * @param globals
 */
export function mockXHR(globals) {
  const response = {
    status: 200,
    data: '',
    headers: '',
  };

  const mockOpen = jest.fn();
  const mockSend = jest.fn(async function (data) {
    await sleep(10);
    this.readyState = 4;
    this.status = response.status ?? 200;
    this.response = response.data || '';
    this.onreadystatechange?.();
    await sleep(10);
    this.onload?.();
  });
  const mockSetRequestHeader = jest.fn();
  globals.FormData = window.FormData;
  globals.XMLHttpRequest = class {
    open = mockOpen;
    send = mockSend;
    setRequestHeader = mockSetRequestHeader;
    getAllResponseHeaders = () => response.headers || '';
  };

  function setResponse(status = 200, data?: any, headers?: Record<string, string>) {
    response.status = status;
    response.data = data;
    if (headers) {
      response.headers = Object.keys(headers)
        .map((key) => {
          const value = headers[key];
          return `${key}: ${value}`;
        })
        .join('\r\n');
    } else {
      response.headers = '';
    }
  }

  return {
    mockOpen,
    mockSend,
    mockSetRequestHeader,
    setResponse,
  };
}

/**
 * 模拟用户选择媒体文件
 */
export function mockUserChooseMedia() {
  const target = new DOMUtil('input[type="file"]');
  target.trigger('change', '');
  // eslint-disable-next-line @typescript-eslint/dot-notation
  target['files'] = [Buffer.alloc(100)];
  return target;
}

/**
 * 模拟 Math.random 方法，返回固定的值
 * @param callback
 */
export async function mockRandomWrapper(callback: () => (Promise<void> | void)) {
  const originalRandom = Math.random;

  let _r = 0;
  Math.random = () => (++_r) / 1e4; // mock random
  await callback();

  // eslint-disable-next-line require-atomic-updates
  Math.random = originalRandom;
}
