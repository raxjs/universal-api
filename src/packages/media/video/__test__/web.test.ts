import { testWebAPI } from '@/utils/__test__/util';
import { mockRandomWrapper, mockUserChooseMedia } from '@/utils/__test__/bom';

testWebAPI('video', async (globals) => {
  await mockRandomWrapper(async () => {
    const { chooseMedia, chooseVideo } = require('../src/index');

    chooseMedia({});
    expect(document.body).toMatchSnapshot();
    await mockUserChooseMedia();
    expect(document.body).toMatchSnapshot();

    chooseVideo({});
    expect(document.body).toMatchSnapshot();
    await mockUserChooseMedia();
    expect(document.body).toMatchSnapshot();
  });
});
