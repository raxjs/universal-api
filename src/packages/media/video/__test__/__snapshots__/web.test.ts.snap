// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Platform API: video Test container: web 1`] = `
<body>
  <input
    id="input-10000"
    multiple="multiple"
    name="file"
    style="display: none;"
    type="file"
  />
</body>
`;

exports[`Test Platform API: video Test container: web 2`] = `<body />`;

exports[`Test Platform API: video Test container: web 3`] = `
<body>
  <input
    accept="video/*"
    id="input-20000"
    name="file"
    style="display: none;"
    type="file"
  />
</body>
`;

exports[`Test Platform API: video Test container: web 4`] = `<body />`;
